packages/ember-toucan-core/src/components/form/controls/combobox.ts

```diff
@@ -0,0 +1,537 @@
+   * We apply different styles to the input tag based on our current state.
+   */
+  get styles() {
+    if (this.args.isDisabled) {
+      return 'shadow-focusable-outline bg-overlay-1 text-disabled pointer-events-none placeholder:text-disabled';
+    }
+
+    if (this.args.isReadOnly) {
+      return 'focus:shadow-focus-outline bg-surface-xl shadow-read-only-outline text-titles-and-attributes placeholder:text-titles-and-attributes';
+    }
+
+    if (this.args.hasError) {
+      return 'shadow-error-outline focus:shadow-error-focus-outline bg-overlay-1 text-titles-and-attributes';
+    }
+
+    return 'shadow-focusable-outline focus:shadow-focus-outline bg-overlay-1 text-titles-and-attributes';
+  }
+
+  /**
+   * The options to render inside of the popover list.
+   */
+  get options() {
+    return this.filteredOptions || this.args?.options;
+  }
+
+  /**
+   * The internal currently selected item.
+   */
+  get selected(): string | undefined {
+    let { optionKey, selected } = this.args;
+
+    if (!selected) {
+      return undefined;
+    }
+
+    return (
+      typeof selected === 'object' && optionKey ? selected[optionKey] : selected
+    ) as string;
+  }
+
+  /**
+   * Attempts to scroll the active or newly highlighted item into view for the user.
+   */
+  #scrollActiveOptionIntoView(alignToTop?: boolean) {
+    assert('`this.activeIndex` cannot be `null`', this.activeIndex !== null);
+
+    const optionsElements = document.querySelectorAll(
+      `#${this.popoverId} ${optionComponentSelector}`
+    );
+    const optionElement = optionsElements[this.activeIndex];
+
+    if (!optionElement) {
+      return;
+    }
+
+    optionElement.scrollIntoView(alignToTop);
+  }
+
+  @action
+  closePopover() {
+    this.isPopoverOpen = false;
+  }
+
+  @action
+  noop() {
+    // eslint-disable @typescript-eslint/no-empty-function
+  }
+
+  /**
+   * Action called when a new item is selected. Ultimately calls the provided `@onChange` with the newly selected item.
+   */
+  @action
+  onChange() {
+    assert('`this.activeIndex` cannot be `null`', this.activeIndex !== null);
+
+    assert(
+      '`this.args.options` cannot be  `undefined`',
+      this.args.options !== undefined
+    );
+
+    let { optionKey, onChange } = this.args;
+
+    this.closePopover();
+
+    // This shouldn't be possible, but to satisfy TS
+    if (!this.options) {
+      onChange?.(null);
+
+      return;
+    }
+
+    // This shouldn't be possible, but to satisfy TS
+    if (this.activeIndex === null) {
+      onChange?.(null);
+
+      return;
+    }
+
+    let selectedOption = this.options[this.activeIndex];
+
+    if (typeof selectedOption === 'string') {
+      this.inputValue = selectedOption;
+    }
+
+    if (selectedOption && typeof selectedOption === 'object' && optionKey) {
+      let option = (selectedOption as Record<string, string>)[optionKey];
+
+      this.inputValue = option;
+    }
+
+    this.args.onChange?.(this.options[this.activeIndex]);
```

Comment:
Normally the user would set whatever gets passed to `@selected` to this new option, but what happens when that does not happen? Maybe there are reasons not to update, like it's an invalid value or something.

I think what happens now is that `inputValue` represents the new option, while the state of `@selected` is still the old one. Which would be inconsistent state, right? 

I think that wouldn't happen if `inputValue` was derived state, as suggested above...

Reply:
This resonated with me at first, but now I find myself questioning it—for better or worse 😆. 

One thing is I'm struggling to imagine a case where a consumer would show an invalid value. 

I also wonder how it would affect the user if the consumer chose not to set `@selected` to the selected value. If the user clicks, say, "Blue" and `@onChange` is called but the consumer decides not to update `@selected`, how should the UI respond? As is, the popover will be closed. But is that the right behavior? How should we let the user know that what he selected is invalid? And if we derive `this.inputValue`, what should its value be?

Part of me thinks that, to prevent all of this, the component shouldn't offer a `@selected` argument and should handle it internally. We could instead offer an `@initialSelected` argument, letting the consumer specify an initially selected state.

Reply:
@simonihmig: So you're up to date, @ynotdraw and I met up to discuss this. 

We realized that my suggestion of `@initialSelected` won't fly if the consumer decides to change `@options` after initial render. There's nothing to stop the consumer from changing `@options` to an entirely different set of options after initial render. And, if they do that, they may also want to update `@initialSelected`. But they won't be able to given we'd only be assigned it to `this.selected` on render. So we took `@initialSelected` off the table.

We also mulled over my user experience questions above and your initial concern about `this.inputValue`. In the end, we decided that we're okay with possibility of `this.inputValue` diverging from `@selected` in the case where the consumer decides not to update `@selected` after `@onChange`. We're okay with this in part because we couldn't imagine a reason why the consumer wouldn't want to update `@selected` and in part of because of the user experience complications I brought up above. 

@ynotdraw brought up another really good point: most consumers (we think) will consume Autocomplete (née Combobox) via the Form component. That means that most consumers won't be setting `@selected` themselves. Form will set it for them.

Let us know what you think? If you can think of a case where the consumer wouldn't want to set `@selected` after `@onChange`, we'd love to here it as well.